File: xor32.pas
unit XOR32;
//XOR32
//Author: domasz
//Version: 0.1 (2022-11-17)
//Licence: MIT

interface

uses SysUtils, HasherBase;

type THasherXOR32 = class(THasherbase)
  private
    FHash: Cardinal;
  public
    constructor Create; override;
    procedure Update(Msg: PByte; Length: Integer); override;
    function Final: String; override;
end;

implementation

constructor THasherXOR32.Create;
begin
  inherited Create;
  Check := '0C04043D';
  FHash := 0;
end;

procedure THasherXOR32.Update(Msg: PByte; Length: Integer);
var i: Integer;
    L: Byte;
begin
  L := 0;
  for i:=0 to Length-1 do begin
    FHash := FHash xor (Msg^ shl L);

    Inc(L, 8);
    if L = 32 then L := 0;

    Inc(Msg);
  end;   
end;

function THasherXOR32.Final: String;
begin
  Result := IntToHex(FHash, 8);
end;

initialization
  HasherList.RegisterHasher('XOR32', THasherXOR32);

end.

File: PHP/Algos/xor32.php
<?php
//XOR-32
//Author: domasz
//Version: 0.1 (2022-11-19)
//Licence: MIT

require_once 'HasherBase.php';

class HasherXOR32 extends HasherBase
{
	protected $FHash = 0;

	public function __construct() 
	{
		$this->Check = '0C04043D';
		$this->FHash = 0;
	}

	public function Update($Msg, $Length)
	{
		$L = 0;
		for ($i=0; $i<$Length; $i++)
		{
			$this->FHash = $this->FHash ^ (ord($Msg[$i]) << $L);
			
			$L +=8;
			if ($L == 32) $L = 0;		
		}
	}

	public function Finish()
	{
		return sprintf('%08X', $this->FHash);
	}
}

$HasherList->RegisterHasher('XOR-32', 'HasherXOR32');